src/core/thread/mesh_forwarder_ftd.cpp

```diff
@@ -184,36 +184,69 @@ void MeshForwarder::HandleResolved(const Ip6::Address &aEid, otError aError)
 
 otError MeshForwarder::EvictMessage(uint8_t aPriority)
 {
-    otError  error = OT_ERROR_NOT_FOUND;
-    Message *message;
+    otError        error    = OT_ERROR_NOT_FOUND;
+    PriorityQueue *queues[] = {&mResolvingQueue, &mSendQueue};
+    Message *      evict    = NULL;
 
-    VerifyOrExit((message = mSendQueue.GetTail()) != NULL);
+    // search for a lower priority message to evict (choose lowest priority message among all queues)
+    for (uint8_t index = 0; index < OT_ARRAY_LENGTH(queues); index++)
+    {
+        for (uint8_t priority = 0; priority < aPriority; priority++)
+        {
+            for (Message *message = queues[index]->GetHeadForPriority(priority); message; message = message->GetNext())
+            {
+                if (message->GetPriority() != priority)
+                {
+                    break;
+                }
+
+                if (message->GetDoNotEvict())
+                {
+                    continue;
+                }
```

Comment:
Can we change the two layer for loop to one like:
```suggestion
    for (Message *message = queues[index]->GetHead(); message && message->GetPriority() < aPriority; message = message->GetNext())
    {
        if (message->GetDoNotEvict())
        {
            continue;
        }
```
I checked some code of `PriorityQueue`:
```
Message *PriorityQueue::GetHeadForPriority(uint8_t aPriority) const
{
    Message *head;
    Message *previousTail;

    if (mTails[aPriority] != NULL)
    {
        previousTail = FindFirstNonNullTail(PrevPriority(aPriority));

        OT_ASSERT(previousTail != NULL);

        head = previousTail->Next();
    }
    else
    {
        head = NULL;
    }

    return head;
}
```
So it seems that the tail of one priority links to the head of next priority. Then I think we can traverse it from low priority to high priority in this way, a little simpler.

Reply:
I don't think that works.

`PriorityQueue::GetHead()` returns the highest priority message. Instead, we want to start with the lowest priority messages.

Note that the `PriorityQueue` implementation relies on a circular linked list:
https://github.com/openthread/openthread/blob/64e03b834741d025aa575f46f692225869031b4d/src/core/common/message.cpp#L799-L802
So it finds the tail of the lowest priority message, then returns the next message, which should be the highest priority message (due to the circularly linked list).

Thoughts?

Reply:
Ok. That's true.